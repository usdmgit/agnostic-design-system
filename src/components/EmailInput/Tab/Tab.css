@import '../../styles/theme/index.css';
:root {
  --tab-container-display: flex;
  --tab-container-flex-flow: column;
  --tab-container-width: auto;

  --tab-menu-container-display: flex;
  --tab-menu-container-flex-flow: row;
  --tab-menu-container-align-items: center;

  --tab-menu-button-height: auto;
  --tab-menu-button-border-width-active: 0 0 2px 0;
  --tab-menu-button-border-width-focus: 0 0 2px 0;
  --tab-menu-button-border-width-hover: 0 0 2px 0;
}

.tab-container {
  display: var(--tab-container-display);
  flex-flow: var(--tab-container-flex-flow);
  width: var(--tab-container-width);
  margin: var(--tab-container-margin);
}

.tab-menu-container {
  align-items: var(--tab-menu-container-align-items);
  display: var(--tab-menu-container-display);
  flex-flow: var(--tab-menu-container-flex-flow);
}

.tab-menu-button {
  --button-color: var(--structure-color-600);
  --button-background-color: transparent;
  --button-color-hover: var(--structure-color-600);
  --button-background-color-hover: transparent;
  --button-border-focus: none;
  --button-border-hover: none;
  --button-border-active: none;
  --button-border-radius: none;
  height: var(--tab-menu-button-height);
}

.tab-menu-button:focus {
  border-width: var(--tab-menu-button-border-width-focus);
}

.tab-menu-button:hover {
  border-width: var(--tab-menu-button-border-width-hover);
}

.tab-menu-button-active {
  color: var(--button-color-active);
  background-color: var(--button-background-color-active);
  border: var(--button-border-active);
  border-width: var(--tab-menu-button-border-width-active);
  --button-color-focus: var(--button-color-active);
  --button-background-color-focus: var(--button-background-color-active);
  --button-border-focus: var(--button-border-active);
  --button-color-hover: var(--button-color-active);
  --button-background-color-hover: var(--button-background-color-active);
  --button-border-hover: var(--button-border-active);
}
