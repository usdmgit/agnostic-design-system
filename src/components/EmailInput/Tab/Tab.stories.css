@import '../../styles/theme/index.css';

.menu-button-active {
  --button-color-active: var(--primary-color-700);
  --button-background-color-active: transparent;
  --button-border-active: 2px solid var(--primary-color-700);
}

.component-container {
  border: 1px solid var(--primary-color-200);
  padding: 20px;
  height: 200px;
  width: 250px;
}
.component-title {
  text-align: center;
}
