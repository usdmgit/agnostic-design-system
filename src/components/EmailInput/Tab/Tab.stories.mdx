# Tab Component

## Type: Simple

## Tab Props

|        Prop        |                        Type                        |             Description              | Default | Mandatory |
| :----------------: | :------------------------------------------------: | :----------------------------------: | :-----: | :-------: |
|     components     |                 React.ReactNode[]                  |    Accepts a list of components.     |    -    |   true    |
|     menuItems      | { title: string; active?: boolean; key: string }[] | Accepts a list of Menu Item objects. |    -    |   true    |
| variablesClassName |                       string                       |   Accepts custom CSS class names.    |    -    |   false   |

## Getting Started

```javascript
import React from 'react';
import { Tab } from '@codelittinc/agnostic-design-system';

const SimpleTab = () => {
  return (
    <Tab
      menuItems={[
        { title: 'Test Active', active: true, key: '1' },
        { title: 'Test Inactive', key: '2' }
      ]}
      components={[
        <>
          <div>Active</div>
        </>,
        <>
          <div>Inactive</div>
        </>
      ]}
    />
  );
};

export default SimpleTab;
```

## Variables available

|      Property      |    Attribute     |             State              |
| :----------------: | :--------------: | :----------------------------: |
|       button       |   align-items    |                                |
|       button       |  align-content   |                                |
|       button       | background-color | hover, focus, active, disabled |
|       button       |      border      | hover, focus, active, disabled |
|       button       |   border-width   |      hover, focus, active      |
|       button       |  border-radius   |                                |
|       button       |    box-shadow    | hover, focus, active, disabled |
|       button       |      color       | hover, focus, active, disabled |
|       button       |      cursor      | hover, focus, active, disabled |
|       button       |     display      |                                |
|       button       |  flex-direction  |                                |
|       button       |    flex-wrap     |                                |
|       button       |   font-family    |                                |
|       button       |    font-size     |                                |
|       button       |   font-weight    |                                |
|       button       |      height      |                                |
|       button       | justify-content  |                                |
|       button       |   line-height    |                                |
|       button       |      margin      |                                |
|       button       |     padding      |                                |
|       button       |      width       |                                |
|     container      |     display      |                                |
|     container      |    flex-flow     |                                |
|     container      |      width       |                                |
|     container      |      margin      |                                |
|   menu-container   |     display      |                                |
|   menu-container   |    flex-flow     |                                |
|   menu-container   |   align-items    |                                |
| menu-button-active |      color       |             hover              |
| menu-button-active | background-color |             hover              |
| menu-button-active |      border      |                                |
| menu-button-active |   border-width   |                                |
|    menu-button     |      height      |                                |
