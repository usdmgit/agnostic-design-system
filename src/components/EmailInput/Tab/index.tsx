import React, { useState } from 'react';
import classNames from 'classnames';
import styles from './Tab.css';
import Button from '@/components/Button';

type MenuItem = {
  title: string | React.ReactNode;
  active?: boolean;
  key: string;
};

interface Props {
  components: React.ReactNode[];
  menuItems: MenuItem[];
  variablesClassName?: string;
}

const Tab = (props: Props) => {
  const { menuItems, components, variablesClassName } = props;
  if (menuItems.length === 0) {
    return <></>;
  }

  const defaultKey = menuItems.find(item => item.active)?.key;
  const [activeKey, setActiveKey] = useState(defaultKey || menuItems[0].key);

  if (menuItems.length !== components.length) {
    return <>It is necessary to provide the same amount of menu items and components.</>;
  }

  return (
    <div className={classNames(styles['tab-container'], variablesClassName)}>
      <div className={classNames(styles['tab-menu-container'], variablesClassName)}>
        {menuItems.map((item, index) => {
          return (
            <Button
              key={`${item.title}-${index}`}
              variablesClassName={classNames(
                styles['tab-menu-button'],
                activeKey === item.key && styles['tab-menu-button-active'],
                variablesClassName
              )}
              text={item.title}
              onClick={() => setActiveKey(item.key)}
            />
          );
        })}
      </div>
      <div>
        {components.map((component, index) => {
          return <div key={index}>{menuItems[index].key === activeKey && component}</div>;
        })}
      </div>
    </div>
  );
};

export default Tab;
