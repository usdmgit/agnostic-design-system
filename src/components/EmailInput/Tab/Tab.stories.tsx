import React from 'react';
import Tab from '@/components/Tab';
import List from '@/components/List';
import { Meta } from '@storybook/react';
import mdx from './Tab.stories.mdx';
import classNames from 'classnames';
import styles from './Tab.stories.css';

export default {
  title: 'Components/Tab',
  component: Tab,
  parameters: {
    docs: {
      page: mdx
    }
  }
} as Meta;

const LIST1 = [
  {
    label: 'Item 1 Tab 1',
    value: 'vlabel'
  },
  {
    label: 'Item 2 Tab 1',
    value: 'vlabel1'
  }
];

const LIST2 = [
  {
    label: 'Item 1 Tab 2',
    value: 'vlabel'
  },
  {
    label: 'Item 2 Tab 2',
    value: 'vlabel1'
  }
];

const LIST3 = [
  {
    label: 'Item 1 Tab 3',
    value: 'vlabel'
  },
  {
    label: 'Item 2 Tab 3',
    value: 'vlabel1'
  }
];

const Template = args => {
  return <Tab {...args} variablesClassName={classNames(styles['menu-button-active'])} />;
};

export const Default = Template.bind({});
Default.args = {
  menuItems: [
    { title: 'Tab 1', active: true, key: '1' },
    { title: 'Tab 2', key: '2' }
  ],
  components: [
    <>
      <div className={classNames(styles['component-container'])}>
        <h2 className={classNames(styles['component-title'])}>Tab 1 Sample Content</h2>
        <List
          options={LIST1}
          id='list'
          getItemValue={item => item.value}
          getItemLabel={item => item.label}
          getItemKey={item => item.value}
          onChange={() => {}}
        />
      </div>
    </>,
    <>
      <div className={classNames(styles['component-container'])}>
        <h2 className={classNames(styles['component-title'])}>Tab 2 Sample Content</h2>
        <List
          options={LIST2}
          id='list'
          getItemValue={item => item.value}
          getItemLabel={item => item.label}
          getItemKey={item => item.value}
          onChange={() => {}}
        />
      </div>
    </>
  ]
};

const renderTitle = text => {
  return (
    <div>
      Tab
      <span>{text}</span>
    </div>
  );
};

export const TabWithHTMLElementTitle = Template.bind({});
TabWithHTMLElementTitle.args = {
  menuItems: [
    { title: renderTitle('1'), active: true, key: '1' },
    { title: renderTitle('2'), key: '2' },
    { title: 'Tab 3', key: '3' }
  ],
  components: [
    <>
      <div className={classNames(styles['component-container'])}>
        <h2 className={classNames(styles['component-title'])}>Tab 1 Sample Content</h2>
        <List
          options={LIST1}
          id='list'
          getItemValue={item => item.value}
          getItemLabel={item => item.label}
          getItemKey={item => item.value}
          onChange={() => {}}
        />
      </div>
    </>,
    <>
      <div className={classNames(styles['component-container'])}>
        <h2 className={classNames(styles['component-title'])}>Tab 2 Sample Content</h2>
        <List
          options={LIST2}
          id='list'
          getItemValue={item => item.value}
          getItemLabel={item => item.label}
          getItemKey={item => item.value}
          onChange={() => {}}
        />
      </div>
    </>,
    <>
      <div className={classNames(styles['component-container'])}>
        <h2 className={classNames(styles['component-title'])}>Tab 3 Sample Content</h2>
        <List
          options={LIST3}
          id='list'
          getItemValue={item => item.value}
          getItemLabel={item => item.label}
          getItemKey={item => item.value}
          onChange={() => {}}
        />
      </div>
    </>
  ]
};
