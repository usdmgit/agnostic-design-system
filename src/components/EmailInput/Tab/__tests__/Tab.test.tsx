import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import Tab from '@/components/Tab';
import { mount } from 'enzyme';

describe('Tab', () => {
  it('renders correctly', () => {
    const wrapper = mount(
      <Tab
        menuItems={[
          { title: 'Test Active', active: true, key: '1' },
          { title: 'Test Inactive', key: '2' }
        ]}
        components={[
          <>
            <div>active</div>
          </>,
          <>
            <div>inactive</div>
          </>
        ]}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  describe('Clicking on an inactive menu button', () => {
    it('opens the corresponding container', () => {
      const { getByText, queryByText } = render(
        <Tab
          menuItems={[
            { title: 'first', active: true, key: '1' },
            { title: 'second', key: '2' }
          ]}
          components={[
            <>
              <div>First Component</div>
            </>,
            <>
              <div>Second Component</div>
            </>
          ]}
        />
      );

      expect(queryByText('First Component')).toBeInTheDocument();
      expect(queryByText('Second Component')).not.toBeInTheDocument();

      fireEvent.click(getByText('second'));

      expect(queryByText('First Component')).not.toBeInTheDocument();
      expect(queryByText('Second Component')).toBeInTheDocument();
    });
  });
});
