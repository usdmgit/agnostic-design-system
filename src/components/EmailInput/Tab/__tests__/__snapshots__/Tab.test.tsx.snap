// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab renders correctly 1`] = `
<Tab
  components={
    Array [
      <React.Fragment>
        <div>
          active
        </div>
      </React.Fragment>,
      <React.Fragment>
        <div>
          inactive
        </div>
      </React.Fragment>,
    ]
  }
  menuItems={
    Array [
      Object {
        "active": true,
        "key": "1",
        "title": "Test Active",
      },
      Object {
        "key": "2",
        "title": "Test Inactive",
      },
    ]
  }
>
  <div
    className="tab-container"
  >
    <div
      className="tab-menu-container"
    >
      <ForwardRef
        key="Test Active-0"
        onClick={[Function]}
        text="Test Active"
        variablesClassName="tab-menu-button tab-menu-button-active"
      >
        <ForwardRef
          onClick={[Function]}
          text="Test Active"
          variablesClassName="tab-menu-button tab-menu-button-active"
        >
          <ForwardRef
            className="button button--neutral"
            disabled={false}
            fixed={false}
            iconPosition="right"
            onClick={[Function]}
            size="large"
            text="Test Active"
            variablesClassName="tab-menu-button tab-menu-button-active"
          >
            <button
              className="button button--neutral button button--large tab-menu-button tab-menu-button-active"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              <span
                className="button-text"
              >
                Test Active
              </span>
            </button>
          </ForwardRef>
        </ForwardRef>
      </ForwardRef>
      <ForwardRef
        key="Test Inactive-1"
        onClick={[Function]}
        text="Test Inactive"
        variablesClassName="tab-menu-button"
      >
        <ForwardRef
          onClick={[Function]}
          text="Test Inactive"
          variablesClassName="tab-menu-button"
        >
          <ForwardRef
            className="button button--neutral"
            disabled={false}
            fixed={false}
            iconPosition="right"
            onClick={[Function]}
            size="large"
            text="Test Inactive"
            variablesClassName="tab-menu-button"
          >
            <button
              className="button button--neutral button button--large tab-menu-button"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              <span
                className="button-text"
              >
                Test Inactive
              </span>
            </button>
          </ForwardRef>
        </ForwardRef>
      </ForwardRef>
    </div>
    <div>
      <div
        key="0"
      >
        <div>
          active
        </div>
      </div>
      <div
        key="1"
      />
    </div>
  </div>
</Tab>
`;
